add MobileNetV2 model

hollance opened the pull request:
# What does this PR do?

Adds MobileNetV2 to the Transformers library.

This includes an image classification head and a basic DeepLabV3+ semantic segmentation head.

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes # (issue)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- albert, bert, xlm: @LysandreJik
- blenderbot, bart, marian, pegasus, encoderdecoder,  t5: @patrickvonplaten, @patil-suraj
- longformer, reformer, transfoxl, xlnet: @patrickvonplaten
- fsmt: @stas00
- funnel: @sgugger
- gpt2: @patrickvonplaten, @LysandreJik
- rag: @patrickvonplaten, @lhoestq
- tensorflow: @LysandreJik

Library:

- benchmarks: @patrickvonplaten
- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam
- text generation: @patrickvonplaten
- tokenizers: @n1t0, @LysandreJik
- trainer: @sgugger
- pipelines: @LysandreJik

Documentation: @sgugger

HF projects:

- datasets: [different repo](https://github.com/huggingface/datasets)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Examples:

- maintained examples (not research project or legacy): @sgugger, @patil-suraj
- research_projects/bert-loses-patience: @JetRunner
- research_projects/distillation: @VictorSanh

 -->

hollance committed:
add model files etc for MobileNetV2

hollance committed:
rename files for MobileNetV1

hollance committed:
initial implementation of MobileNetV1

hollance committed:
fix conversion script

hollance committed:
cleanup

hollance committed:
write docs

hollance committed:
tweaks

hollance committed:
fix conversion script

hollance committed:
extract hidden states

hollance committed:
fix test cases

hollance committed:
make fixup

hollance committed:
fixup it all

hollance committed:
rename V1 to V2

hollance committed:
fix checkpoints

hollance committed:
fixup

hollance committed:
implement first block + weight conversion

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_17845). All of your documentation changes will be reflected on that endpoint.

hollance committed:
add remaining layers

hollance committed:
add output stride and dilation

hollance committed:
fixup

hollance committed:
add tests

hollance committed:
add deeplabv3+ head

hollance committed:
a bit of fixup

hollance committed:
finish deeplab conversion

hollance committed:
add link to doc

hollance committed:
fix issue with JIT trace

in_height and in_width would be Tensor objects during JIT trace, which caused Core ML conversion to fail on the remainder op. By making them ints, the result of the padding calculation becomes a constant value.

sgugger reviewed (approved):
Very clean, thanks a lot for the new model!

hollance committed:
cleanup

hollance committed:
fix order of models

hollance commented on tests/models/mobilenet_v2/test_modeling_mobilenet_v2.py:
```diff
@@ -0,0 +1,346 @@
+                self.batch_size,
+                self.last_hidden_size,
+            ),
+        )
```
OK, done.

hollance committed:
fix rebase error

hollance committed:
remove main from doc link

hollance committed:
add image processor

hollance committed:
remove old feature extractor

hollance committed:
fix converter + other issues

hollance committed:
fixup

hollance committed:
fix unit test

hollance committed:
add to onnx tests (but these appear broken now)

hollance committed:
add post_process_semantic_segmentation

hollance committed:
use google org

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_17845). All of your documentation changes will be reflected on that endpoint.

sgugger commented on src/transformers/models/mobilenet_v2/image_processing_mobilenet_v2.py:
```diff
@@ -0,0 +1,381 @@
+from ...utils import logging
+
+
+if is_vision_available():
+    import PIL
```
Apparently this is not used according to `flake8`.

hollance replied:
yeah, I was too enthusiastic copy-pasting stuff :-)

sgugger commented on src/transformers/models/mobilenet_v2/image_processing_mobilenet_v2.py:
```diff
@@ -0,0 +1,381 @@
+
+    def post_process_semantic_segmentation(self, outputs, target_sizes: List[Tuple] = None):
+        """
+        Args:
+        Converts the output of [`MobileNetV2ForSemanticSegmentation`] into semantic segmentation maps. Only supports
+        PyTorch.
+            outputs ([`MobileNetV2ForSemanticSegmentation`]):
+                Raw outputs of the model.
+            target_sizes (`List[Tuple]`, *optional*):
+                A list of length `batch_size`, where each item is a `Tuple[int, int]` corresponding to the requested
+                final size (height, width) of each prediction. If left to None, predictions will not be resized.
```
Args should be below the docstring.

sgugger reviewed (commented):
A couple of final nits before we merge this :-)

hollance committed:
remove unused imports

hollance committed:
move args

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_17845). All of your documentation changes will be reflected on that endpoint.

hollance committed:
replace weird assert

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_17845). All of your documentation changes will be reflected on that endpoint.

sgugger commented:
Good to merge @hollance ?

hollance commented:
@sgugger It seems to fail some tests but they don't look they're from my changes? If you're OK with that test failing then I think this is ready to get merged.

sgugger commented:
Failure is a flaky test, unrelated to this PR.
